### PR TITLE
subprocess improvements

### LINKED ISSUE #485: check subprocess id to see if server is running rather than timing out
The timeout for a server to start running will need to be a long time for larger models, better to just check the process ID then wait for the server to respond (with a really long timeout) rather than relying on the timeout by itself.

### PR DESCRIPTION
**Pull request by BruceMacD**
- increase start-up timeout
- when runner fails to start fail rather than timing out
- try runners in order rather than choosing 1 runner
- embed metal runner in metal dir rather than gpu
- refactor logging and error messages

resolves #485

### CODE REVIEW

File: llm/llama.go [hunk 1]
```diff
@@ -371,35 +469,40 @@ func newLlama(model string, adapters []string, runner ModelRunner, opts api.Opti
...
+			return fmt.Errorf("llama runner did not start within alloted time, retrying")
+		}
+
+		// check if the server process has terminated
+		select {
+		case <-llm.Exited:
+			return fmt.Errorf("llama runner process has terminated")
+		default:
+			// command is still running
 		}
```
Comment (mxyng) [inline llm/llama.go:493]: This isn't necessary if you check `llm.Cmd.ProcessState`

```suggestion
		if llm.Cmd.ProcessState != nil && llm.Cmd.ProcessState.Exited() {
			return fmt.Errorf("llama runner process has terminated")
		}
```

File: llm/llama.go [hunk 2]
```diff
@@ -354,14 +438,28 @@ func newLlama(model string, adapters []string, runner ModelRunner, opts api.Opti
...
+		log.Print("starting llama runner")
 		if err := llm.Cmd.Start(); err != nil {
-			log.Printf("error starting the external llama.cpp server: %v", err)
+			log.Printf("error starting the external llama runner: %v", err)
 			continue
 		}
 
+		// monitor the command, it is blocking, so if it exits we need to capture that
+		go func() {
+			err := llm.Cmd.Wait() // this will block until the command exits
```
Comment (mxyng) [inline llm/llama.go:449]: I'm not a fan of this structurally but I can't think of an alternative. That suggests this code needs to be refactored, perhaps separating the subprocess runner from the llm

Comment (BruceMacD) [reply llm/llama.go:380]: I agree with this, lets do it in a follow-up change to keep things clean in my opinion.

File: llm/llama.go [hunk 3]
```diff
@@ -59,104 +51,128 @@ func cudaVersion() (int, error) {
...
-		srcFile, err := llamaCppEmbed.Open(srcPath)
+		srcFile, err := llamaCppEmbed.Open(f)
 		if err != nil {
-			log.Fatalf("read llama.cpp %s: %v", f, err)
+			log.Fatalf("read llama runner %s: %v", f, err)
 		}
 		defer srcFile.Close()
 
+		// create the directory in case it does not exist
+		destPath := filepath.Join(tmpDir, strings.TrimPrefix(f, buildPath))
```
Comment (mxyng) [inline llm/llama.go:126]: Is it necessary to preserve part of the path? I would imagine `filepath.Dir` would be sufficient for most cases.

File: llm/llama.go [hunk 4]
```diff
@@ -59,104 +51,128 @@ func cudaVersion() (int, error) {
...
-		log.Printf("llama.cpp %s executable not found, falling back to cpu", runnerType)
 	}
 
 	// copy the files locally to run the llama.cpp server
+	tmpDir, err := os.MkdirTemp("", "llama-*")
+	if err != nil {
+		log.Fatalf("load llama runner: failed to create temp dir: %v", err)
+	}
+	runnerAvailable := false // if no runner files are found in the embed, this flag will cause a fast fail
 	for _, f := range files {
```
Comment (mxyng) [inline llm/llama.go:111]: Instead of static files, perhaps use `fs.Glob` and copy everything

Comment (BruceMacD) [reply llm/llama.go]: Nice, I did this and it made things simpler in another pass

**Review by mxyng — APPROVED** [review]

**Pull request merged**